### PR TITLE
Taxon patch

### LINKED ISSUE #17: The taxon function has no method to pass a taxon ID directly
Currently, `taxon` performs only text search, but sometimes the ID of the taxon is known (for example, when we want the genus of a given species). This would require a second method.

The `taxon` function is defined here: https://github.com/EcoJulia/GBIF.jl/blob/master/src/taxon.jl

The GBIF API endpoint for taxa by ID is here: http://api.gbif.org/v1/species/216

### PR DESCRIPTION
**Pull request by MichielStock**
As requested in Issue #17, I wrote a function to query when the ID of the species is known.

Not 100% finished (e.g. no unit test at the moment).

Ping @tpoisot to take a look.

*@tpoisot edit:* will fix #17 when merged

### CODE REVIEW

File: src/taxon.jl [hunk 1]
```diff
@@ -80,3 +80,92 @@ function taxon(name::String;
...
+
+"""
+**Get information about a taxon at any level using taxonID**
+
+    taxon(id::Int)
+
+This function will look for a taxon by its taxonID in the GBIF
+reference taxonomy.
+
+Optional arguments are
```
Comment (tpoisot) [inline src/taxon.jl]: You can remove pretty much everything, since this function will not accept keyword arguments.

File: src/taxon.jl [hunk 2]
```diff
@@ -80,3 +80,92 @@ function taxon(name::String;
...
+- `strict::Bool=true` -- whether the match should be strict, or fuzzy
+
+- `verbose::Bool=false` -- whether the query should print out information about the reply (not usually useful)
+
+Finally, one can also specify other levels of the taxonomy, using  `kingdom`,
+`phylum`, `class`, `order`, `family`, and `genus`, all of which can either be
+`String` or `Nothing`.
+"""
+function mytaxon(id::Int;
+   rank::Union{Symbol,Nothing}=:SPECIES, strict::Bool=true, verbose::Bool=false,
```
Comment (tpoisot) [inline src/taxon.jl]: Again, everything from here to line 150 can be removed -- in the function with a taxon name, these are used to restrict the search; but as an ID will necessarily point to a single taxon, they are not needed.

File: src/taxon.jl [hunk 3]
```diff
@@ -80,3 +80,92 @@ function taxon(name::String;
...
+      matchtype == "NONE" && throw(ErrorException("No match for taxonID $(id) at level $(rank) -- try with strict=false"))
+      =#
+      return GBIFTaxon(body)
+   else
+      throw(ErrorException("Impossible to retrieve information for taxonID $(id) -- HTML error code $(sp_s_req.status)"))
+   end
+
+end
+
+mytaxon(t::Pair) = taxon(t.second)
```
Comment (tpoisot) [inline src/taxon.jl]: Change to `taxon(t::Pair) = taxon(t.second)`, and add a documentation specifying that this works for `name => id` pairs.

File: src/taxon.jl [hunk 4]
```diff
@@ -80,3 +80,92 @@ function taxon(name::String;
...
+      args["genus"] = String(genus)
+   end
+
+   sp_s_url = gbifurl * "species/$id"
+   sp_s_req = HTTP.get(sp_s_url, query=args)
+   if sp_s_req.status == 200
+      body = JSON.parse(String(sp_s_req.body))
+      body["status"] = "ACCEPTED"
+      body["matchType"] = "EXACT"
+      #=  # no match type here
```
Comment (tpoisot) [inline src/taxon.jl]: Do you want to make these fields optional? There is another place in the package where we don't have a `matchType`, so maybe a change to `GBIFTaxon` is in order.

**Review by tpoisot — CHANGES_REQUESTED** [review]

File: src/taxon.jl (hunk 1 above)
Comment (MichielStock) [reply src/taxon.jl]: Ok, will do, I though adding that stuff still made sense

File: src/taxon.jl (hunk 3 above)
Comment (MichielStock) [reply src/taxon.jl]: Artifact from debugging, will fix!

File: src/taxon.jl (hunk 4 above)
Comment (MichielStock) [reply src/taxon.jl]: I will make these optional in `GBIFTaxon` then.

Comment (MichielStock) [reply src/taxon.jl]: solved in `GBIFTaxon.jl` by default missing values.

```julia
   return GBIFTaxon(
      o["canonicalName"],
      o["scientificName"],
      Symbol(get(o, "status", "none")),  # returns :none as default
      Symbol(get(o, "matchType", "none")),
      r["kingdom"],
      r["phylum"],
      r["class"],
      r["order"],
      r["family"],
      r["genus"],
      r["species"],
      get(o, "confidence", missing),
      o["synonym"]
   )
```

File: src/types/GBIFTaxon.jl [hunk 5]
```diff
@@ -59,16 +59,16 @@ function GBIFTaxon(o::Dict{String, Any})
    return GBIFTaxon(
       o["canonicalName"],
       o["scientificName"],
-      Symbol(o["status"]),
-      Symbol(o["matchType"]),
+      Symbol(get(o, "status", "none")),  # returns :none as default
```
Comment (tpoisot) [inline src/types/GBIFTaxon.jl]: What about allowing `Missing` here too?

**Review by tpoisot — CHANGES_REQUESTED** [review]
Seems good, except for the one comment about allowing `missing` in one more place.

**Comment by tpoisot** [comment]
It's also missing the tests!

**Comment by MichielStock** [comment]
Done!

**Review by tpoisot — APPROVED** [review]

**Comment by mkborregaard** [comment]
@tpoisot I can see you've approved the PR but tests are failing (at least on Travis, seems to be in filtering), and there's still the open comment on `missing`?
(as an aside - should we start using the `testset` functionality? The Travis report is not very informative about where the error is being thrown.

**Comment by MichielStock** [comment]
> @tpoisot I can see you've approved the PR but tests are failing (at least on Travis, seems to be in filtering), and there's still the open comment on `missing`?

I did not change anything in filtering, and there is a similar error in an other branch. This seems like an issue unrelated to my patch.

**Comment by mkborregaard** [comment]
Great - what's your take on the `missing` thing? Did you solve this and I didn't notice or did you disagree?

**Comment by mkborregaard** [comment]
Awesome work @MichielStock !

**Comment by tpoisot** [comment]
> > @tpoisot I can see you've approved the PR but tests are failing (at least on Travis, seems to be in filtering), and there's still the open comment on `missing`?
> 
> I did not change anything in filtering, and there is a similar error in an other branch. This seems like an issue unrelated to my patch.

I will have to see what is happening. I probably have written tests that can fail/pass depending on the state of the most recent data, so I'll need to write something that I know for sure will behave.

**Pull request merged**